src/openms/include/OpenMS/DATASTRUCTURES/ExposedVector.h

```diff
@@ -0,0 +1,223 @@
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow $
+// $Authors: Chris Bielow $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+
+#include <vector>
+
+namespace OpenMS
+{
+
+
+#define EXPOSED_VECTOR_INTERFACE(InnerElement) \
+    using ExpVec = ExposedVector< InnerElement >;                     \
+    using ExpVec::ExposedVector;                                               \
+    using value_type = typename ExpVec::value_type;                            \
+    using iterator = typename ExpVec::iterator;                                \
+    using const_iterator = typename ExpVec::const_iterator;                    \
+    using reverse_iterator = typename ExpVec::reverse_iterator;                \
+    using const_reverse_iterator = typename ExpVec::const_reverse_iterator;    \
+    using size_type = typename ExpVec::size_type;                              \
+    using pointer = typename ExpVec::pointer;                                  \
+    using reference = typename ExpVec::reference;                              \
+    using const_reference = typename ExpVec::const_reference;                  \
+    using difference_type = typename ExpVec::difference_type;                  \
+ 
+
+  /**
+    @brief Makes a vector<VectorElement> available in the derived class and exposed commonly
+           used vector member functions at class level.
+
+    This saves writing repetitive code which forwards commonly used functions of a data member, e.g. 'data_.begin()' 
+    as a member function of the class. Also it makes private inheritance from vector<VectorElement> obsolete.
+    The latter is problematic for many reasons (read up on 'prefer composition over inheritance'). In our case, even linking
+    can be problematic with private inheritance once you require RTTI (which some tools do, e.g. softwipe).
+
+    To fully utilize this class (i.e. access the 'iterator' type), call 
+\code
+    EXPOSED_VECTOR_INTERFACE(VectorElement)
+\endcode
+
+   in your derived class, where @p VectorElement is identical to the template argument of ExposedVector, e.g. 'Feature' for FeatureMap.
+
+
+    @ingroup Datastructures
+  */
+
+  template<class VectorElement>
+  class ExposedVector
+  {
+  public: 
+    template<class T>
+    friend class UniqueIdIndexer; ///< accesses data_ but is not derived from ExposedVector
```

Comment:
the friendship / tieing ExposedVector to UID is a bit of a "design smell". Would a const getExposedVector help here? I think we don't have a proper abstraction anyway.

Reply:
yes, true. I wanted to avoid having to put `friend...` in all classes which use the UniqueIdIndexer.
I added a member `.getData()` now, which allows public access to the data (which users had anyway via all the other functions exposed). It's still a bit of a compromise, but it does not smell badly anymore :)